File: framework/core/BaseAPI.ts
import { APIRequestContext, APIResponse, expect } from '@playwright/test';
import Ajv from 'ajv';
import { HttpMethod } from '../../core/api/types';
import { getNestedProperty } from '../../core/api/variableUtils';

const ajv = new Ajv({ allErrors: true });

export class BaseAPI {
  protected requestContext: APIRequestContext;

  constructor(requestContext: APIRequestContext) {
    this.requestContext = requestContext;
  }

  /** Used by generated API clients and runners that wrap the same Playwright context. */
  public getRequestContext(): APIRequestContext {
    return this.requestContext;
  }

  public async get(url: string, options?: Parameters<APIRequestContext['get']>[1]): Promise<APIResponse> {
    console.log(`[BaseAPI] GET ${url}`);
    return this.requestContext.get(url, options);
  }

  public async post(url: string, data?: unknown, options?: Parameters<APIRequestContext['post']>[1]): Promise<APIResponse> {
    console.log(`[BaseAPI] POST ${url}`);
    return this.requestContext.post(url, { data, ...options });
  }

  public async put(url: string, data?: unknown, options?: Parameters<APIRequestContext['put']>[1]): Promise<APIResponse> {
    console.log(`[BaseAPI] PUT ${url}`);
    return this.requestContext.put(url, { data, ...options });
  }

  public async patch(url: string, data?: unknown, options?: Parameters<APIRequestContext['patch']>[1]): Promise<APIResponse> {
    console.log(`[BaseAPI] PATCH ${url}`);
    return this.requestContext.patch(url, { data, ...options });
  }

  public async delete(url: string, options?: Parameters<APIRequestContext['delete']>[1]): Promise<APIResponse> {
    console.log(`[BaseAPI] DELETE ${url}`);
    return this.requestContext.delete(url, options);
  }

  public async head(url: string, options?: Parameters<APIRequestContext['head']>[1]): Promise<APIResponse> {
    console.log(`[BaseAPI] HEAD ${url}`);
    return this.requestContext.head(url, options);
  }

  public async request(
    method: HttpMethod,
    url: string,
    options?: { headers?: Record<string, string>; data?: unknown }
  ): Promise<APIResponse> {
    console.log(`[BaseAPI] ${method} ${url}`);
    return this.requestContext.fetch(url, { method, ...options });
  }

  public async assertStatus(response: APIResponse, expectedStatus: number): Promise<void> {
    const status = response.status();
    console.log(`[BaseAPI] Assert status ${expectedStatus} (actual: ${status})`);
    expect(status).toBe(expectedStatus);
  }

  public async assertBodyContains(response: APIResponse, text: string): Promise<void> {
    const textBody = await response.text();
    console.log(`[BaseAPI] Assert body contains: "${text}"`);
    expect(textBody).toContain(text);
  }

  public async assertJsonPathExists(json: unknown, jsonPath: string): Promise<void> {
    const value = getNestedProperty(json, jsonPath);
    console.log(`[BaseAPI] Assert JSON path exists: ${jsonPath}`);
    expect(value).not.toBeUndefined();
  }

  public async assertJsonPathEquals(json: unknown, jsonPath: string, expected: unknown): Promise<void> {
    const value = getNestedProperty(json, jsonPath);
    console.log(`[BaseAPI] Assert ${jsonPath} === ${JSON.stringify(expected)}`);
    expect(value).toEqual(expected);
  }

  public extractFromResponse<T = unknown>(json: unknown, jsonPath: string): T | undefined {
    return getNestedProperty(json, jsonPath) as T | undefined;
  }

  public async validateSchema(response: APIResponse, schema: object): Promise<void> {
    console.log(`[BaseAPI] Validating JSON schema...`);
    const json = await response.json();
    const validate = ajv.compile(schema);
    const valid = validate(json);
    if (!valid) {
      const errorText = ajv.errorsText(validate.errors);
      console.error(`[BaseAPI] Schema validation failed:`, errorText);
      throw new Error(`JSON Schema validation failed: ${errorText}`);
    }
    console.log(`[BaseAPI] Schema validated.`);
  }

  public async getJson<T = unknown>(response: APIResponse): Promise<T> {
    return response.json() as Promise<T>;
  }

  public async getText(response: APIResponse): Promise<string> {
    return response.text();
  }

  public async saveResponseBody(response: APIResponse, filePath: string): Promise<void> {
    const fs = await import('fs');
    const body = await response.text();
    fs.writeFileSync(filePath, body, 'utf8');
  }
}

File: framework/core/base_api.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from jsonschema import validate
from playwright.sync_api import APIRequestContext, APIResponse


def get_nested_property(value: Any, path: str) -> Any:
    current = value
    for part in path.replace("[", ".").replace("]", "").split("."):
        if not part:
            continue
        if isinstance(current, list):
            current = current[int(part)]
        elif isinstance(current, dict):
            current = current.get(part)
        else:
            return None
    return current


class BaseAPI:
    def __init__(self, request_context: APIRequestContext) -> None:
        self.request_context = request_context

    def get(self, url: str, **options: Any) -> APIResponse:
        return self.request_context.get(url, **options)

    def post(self, url: str, data: Any = None, **options: Any) -> APIResponse:
        return self.request_context.post(url, data=data, **options)

    def put(self, url: str, data: Any = None, **options: Any) -> APIResponse:
        return self.request_context.put(url, data=data, **options)

    def patch(self, url: str, data: Any = None, **options: Any) -> APIResponse:
        return self.request_context.patch(url, data=data, **options)

    def delete(self, url: str, **options: Any) -> APIResponse:
        return self.request_context.delete(url, **options)

    def head(self, url: str, **options: Any) -> APIResponse:
        return self.request_context.head(url, **options)

    def request(self, method: str, url: str, **options: Any) -> APIResponse:
        return self.request_context.fetch(url, method=method, **options)

    @staticmethod
    def assert_status(response: APIResponse, expected_status: int) -> None:
        assert response.status == expected_status

    @staticmethod
    def assert_body_contains(response: APIResponse, text: str) -> None:
        assert text in response.text()

    @staticmethod
    def assert_json_path_exists(payload: Any, json_path: str) -> None:
        assert get_nested_property(payload, json_path) is not None

    @staticmethod
    def assert_json_path_equals(payload: Any, json_path: str, expected: Any) -> None:
        assert get_nested_property(payload, json_path) == expected

    @staticmethod
    def validate_schema(response: APIResponse, schema: dict[str, Any]) -> None:
        validate(instance=response.json(), schema=schema)

    @staticmethod
    def save_response_body(response: APIResponse, file_path: str | Path) -> None:
        Path(file_path).write_text(response.text(), encoding="utf-8")
